packages/playwright-core/src/server/network.ts

```diff
@@ -43,14 +43,18 @@ export function filterCookies(cookies: channels.NetworkCookie[], urls: string[])
         continue;
       if (!parsedURL.pathname.startsWith(c.path))
         continue;
-      if (parsedURL.protocol !== 'https:' && parsedURL.hostname !== 'localhost' && c.secure)
+      if (parsedURL.protocol !== 'https:' && !isLocalHostname(parsedURL.hostname) && c.secure)
```

Comment:
Since you are changing logic for browsers too, please add a test for the cookies received by the browsers from the server. You can use `proxyServer` fixture, see [this test](https://github.com/microsoft/playwright/blob/a6b4b222ed67515b1694ee751229a4ebfc66a2fe/tests/library/browsercontext-proxy.spec.ts#L55-L65) for an example, you'll just need to use `page.goto()` for navigation. My understanding is that before this change `BrowserContext.cookies('http://subdomain.localhost:3000/')` would not return secure cookies and now it does.